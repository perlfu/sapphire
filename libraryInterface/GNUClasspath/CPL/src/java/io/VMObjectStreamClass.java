/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.io;

import java.lang.reflect.Field;

import org.jikesrvm.classloader.VM_Field;
import org.jikesrvm.classloader.VM_Type;

/**
 * java.io.ObjectStream helper implemented for Jikes RVM.
 */
final class VMObjectStreamClass {

  static boolean hasClassInitializer (Class<?> cls) {
    VM_Type t = java.lang.JikesRVMSupport.getTypeForClass(cls);
    if (t.isClassType()) {
      return t.asClass().getClassInitializerMethod() != null;
    } else {
      return false;
    }
  }

  static void setDoubleNative(Field field, Object obj, double val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setDoubleValueUnchecked(obj, val);
  }

  static void setFloatNative(Field field, Object obj, float val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setFloatValueUnchecked(obj, val);
  }

  static void setLongNative(Field field, Object obj, long val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setLongValueUnchecked(obj, val);
  }

  static void setIntNative(Field field, Object obj, int val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setIntValueUnchecked(obj, val);
  }

  static void setShortNative(Field field, Object obj, short val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setShortValueUnchecked(obj, val);
  }

  static void setCharNative(Field field, Object obj, char val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setCharValueUnchecked(obj, val);
  }

  static void setByteNative(Field field, Object obj, byte val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setByteValueUnchecked(obj, val);
  }

  static void setBooleanNative(Field field, Object obj, boolean val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setBooleanValueUnchecked(obj, val);
  }

  static void setObjectNative(Field field, Object obj, Object val) {
    VM_Field f = java.lang.reflect.JikesRVMSupport.getFieldOf(field);
    f.setObjectValueUnchecked(obj, val);
  }
}
