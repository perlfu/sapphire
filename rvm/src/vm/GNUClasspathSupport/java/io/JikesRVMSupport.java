/*
 * (C) Copyright IBM Corp 2002
 */
//$Id$
package java.io;

import com.ibm.JikesRVM.librarySupport.FileSupport;

/**
 * Library support interface of Jikes RVM
 *
 * $Id$
 *
 * @author Julian Dolby
 */
public class JikesRVMSupport {

    /**
     * Create a <code>FileDescriptor</code> object.
     *
     * @param fd the underlying OS file descriptor
     * @param shared true if the OS file descriptor may possibly be shared with
     *   other processes, false if not
     * @return the new <code>FileDescriptor</code> object
     */
    public static FileDescriptor createFileDescriptor(int unixFd, boolean shared) {
	FileDescriptor javaFd = new FileDescriptor(unixFd);
	FileSupport.onCreateFileDescriptor(javaFd, shared);
	return javaFd;
    }

    /**
     * Modify an existing <code>FileDescriptor</code> object to refer
     * to given OS file descriptor.
     *
     * @param fdObj the <code>FileDescriptor</code> object
     * @param fd the OS file descriptor
     */
    public static void setFd(FileDescriptor fdObj, int fd) {
	fdObj.setNativeFD( fd );
    }

    /**
     * Get the underlying OS file descriptor associated with given
     * <code>FileDescriptor<code> object.
     *
     * @param fdObj the <code>FileDescriptor</code> object
     */
    public static int getFd(FileDescriptor fdObj) {
	return fdObj.getNativeFD();
    }
    
}

