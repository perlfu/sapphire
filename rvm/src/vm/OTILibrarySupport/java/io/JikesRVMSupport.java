/*
 * (C) Copyright IBM Corp. 2002
 */
//$Id$

package java.io;

/**
 * Interface for Jikes RVM to access hidden implementation
 * fields and methods of classes in the <code>java.io</code>
 * package.
 *
 * @author David Hovemeyer
 */
public class JikesRVMSupport {
  /**
   * Create a <code>FileDescriptor</code> object.
   *
   * @param fd the underlying OS file descriptor
   * @param shared true if the OS file descriptor may possibly be shared with
   *   other processes, false if not
   * @return the new <code>FileDescriptor</code> object
   */
  public static FileDescriptor createFileDescriptor(int fd, boolean shared) {
    return new FileDescriptor(fd, shared);
  }

  /**
   * Modify an existing <code>FileDescriptor</code> object to refer
   * to given OS file descriptor.
   *
   * @param fdObj the <code>FileDescriptor</code> object
   * @param fd the OS file descriptor
   */
  public static void setFd(FileDescriptor fdObj, int fd) {
    fdObj.fd = fd;
  }

  /**
   * Get the underlying OS file descriptor associated with given
   * <code>FileDescriptor<code> object.
   *
   * @param fdObj the <code>FileDescriptor</code> object
   */
  public static int getFd(FileDescriptor fdObj) {
    return fdObj.fd;
  }

  /**
   * Add a reference to given <code>FileDescriptor</code> object.
   * Increments the reference count.
   *
   * @param fdObj the <code>FileDescriptor</code> object
   * @return the same object
   */
  public static FileDescriptor ref(FileDescriptor fd) {
    return fd.ref();
  }

  /**
   * Remove a reference to given <code>FileDescriptor</code> object.
   * Decrements the reference count.
   *
   * @param fdObj the <code>FileDescriptor</code> object
   */
  public static void deref(FileDescriptor fd) {
    fd.deref();
  }

  /**
   * Determine whether or not the given <code>FileDescriptor</code> object
   * is dead (meaning that its reference count has become zero).
   *
   * @param fdObj the <code>FileDescriptor</code> object
   */
  public static boolean dead(FileDescriptor fd) {
    return fd.dead();
  }
}
