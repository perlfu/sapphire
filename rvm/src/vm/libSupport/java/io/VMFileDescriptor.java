/*
 * (C) Copyright IBM Corp 2002
 */
// $Id$
package java.io;

import com.ibm.JikesRVM.VM_FileSystem;

/**
 * FileDescriptor provider interface, implemented for Jikes RVM.
 *
 * @author Julian Dolby
 */
final class VMFileDescriptor {

  /**
    *  This method is called in the class initializer to do any require
    * native library initialization.  It is also responsible for initializing
    * the in, out, and err variables.
    */
    static void nativeInit() {

    }

  /**
    * Opens the specified file in the specified mode.  This can be done
    * in one of the specified modes:
    * <ul>
    * <li>r - Read Only
    * <li>rw - Read / Write
    * <li>ra - Read / Write - append to end of file
    * <li>rws - Read / Write - synchronous writes of data/metadata
    * <li>rwd - Read / Write - synchronous writes of data.
    *
    * @param path Name of the file to open
    * @param mode Mode to open
    *
    * @returns The resulting file descriptor for the opened file, or -1
    * on failure (exception also signaled).
    *
    * @exception IOException If an error occurs.
    */
  static long nativeOpen(String path, String mode) throws IOException {
      int fd;
      if ("r".equals(mode))
	  fd = VM_FileSystem.open(path, VM_FileSystem.OPEN_READ);
      else if ("w".equals(mode))
	  fd = VM_FileSystem.open(path, VM_FileSystem.OPEN_WRITE);
      else if ("rw".equals(mode)||"rwd".equals(mode)||"rws".equals(mode))
	  fd = VM_FileSystem.open(path, VM_FileSystem.OPEN_MODIFY);
      else if ("ra".equals(mode)||"a".equals(mode))
	  fd = VM_FileSystem.open(path, VM_FileSystem.OPEN_APPEND);
      else
	  fd = -1;

      if (fd == -1)
	  throw new IOException("Could not create/open file " + path);
      else {
	  VM_FileSystem.onCreateFileDescriptor(fd, false);
	  return fd;
      }
  }

  /**
    * Closes this specified file descriptor
    * 
    * @param fd The native file descriptor to close
    *
    * @returns The return code of the native close command.
    *
    * @exception IOException If an error occurs 
    */    
  static long nativeClose(long fd) throws IOException {
      return VM_FileSystem.close( (int)fd );
  }
 
  /**
    * Writes a single byte to the file
    *
    * @param fd The native file descriptor to write to
    * @param b The byte to write, encoded in the low eight bits
    *
    * @return The return code of the native write command
    *
    * @exception IOException If an error occurs
    */
  static long nativeWriteByte(long fd, int b) throws IOException {
      return VM_FileSystem.writeByte((int)fd, b);
  }

  /**
    * Writes a byte buffer to the file
    *
    * @param fd The native file descriptor to write to
    * @param buf The byte buffer to write from
    * @param int The offset into the buffer to start writing from
    * @param len The number of bytes to write.
    *
    * @return The return code of the native write command
    *
    * @exception IOException If an error occurs
    */
  static long nativeWriteBuf(long fd, byte[] buf, int offset, int len)
      throws IOException
  {
      return VM_FileSystem.writeBytes((int)fd, buf, offset, len);
  }

  /**
    * Reads a single byte from the file
    *
    * @param fd The native file descriptor to read from
    *
    * @return The byte read, in the low eight bits on a long, or -1
    * if end of file
    *
    * @exception IOException If an error occurs
    */
  static int nativeReadByte(long fd) throws IOException {
      return VM_FileSystem.readByte((int)fd);
  }

  /**
    * Reads a buffer of  bytes from the file
    *
    * @param fd The native file descriptor to read from
    * @param buf The buffer to read bytes into
    * @param offset The offset into the buffer to start storing bytes
    * @param len The number of bytes to read.
    *
    * @return The number of bytes read, or -1 if end of file.
    *
    * @exception IOException If an error occurs
    */
  static int nativeReadBuf(long fd, byte[] buf, int offset, int len) 
      throws IOException
  {
      int bytesRead = VM_FileSystem.readBytes((int)fd, buf, offset, len);
      if (bytesRead == 0 && len != 0)
	  return -1; // EOF
      else
	  return bytesRead;
  }

  /**
    * Returns the number of bytes available for reading
    *
    * @param fd The native file descriptor
    *
    * @return The number of bytes available for reading
    *
    * @exception IOException If an error occurs
    */
  static int nativeAvailable(long fd) throws IOException {
      return VM_FileSystem.bytesAvailable( (int)fd );
  }

  /**
    * Method to do a "seek" operation on the file
    * 
    * @param fd The native file descriptor 
    * @param offset The number of bytes to seek
    * @param whence The position to seek from, either
    *    SET (0) for the beginning of the file, CUR (1) for the 
    *    current position or END (2) for the end position.
    * @param stopAtEof <code>true</code> to ensure that there is no
    *    seeking past the end of the file, <code>false</code> otherwise.
    *
    * @return The new file position, or -1 if end of file.
    *
    * @exception IOException If an error occurs
    */
  static long nativeSeek(long fd, long offset, int whence, boolean stopAtEof)
    throws IOException 
  {
      return VM_FileSystem.seek((int)fd, (int)offset, whence);
  }

  /**
    * Returns the current position of the file pointer in the file
    *
    * @param fd The native file descriptor
    *
    * @exception IOException If an error occurs
    */
  static long nativeGetFilePointer(long fd) throws IOException {
      return nativeSeek((int)fd, 0L, FileDescriptor.CUR, false);
  }

  /**
    * Returns the length of the file in bytes
    *
    * @param fd The native file descriptor
    *
    * @return The length of the file in bytes
    *
    * @exception IOException If an error occurs
    */
  static long nativeGetLength(long fd) throws IOException {
      return VM_FileSystem.length( (int)fd );
  }

  /**
    * Sets the length of the file to the specified number of bytes
    * This can result in truncation or extension.
    *
    * @param fd The native file descriptor  
    *
    * @param len The new length of the file
    *
    * @exception IOException If an error occurs
    */
  static void nativeSetLength(long fd, long len) throws IOException {
      VM_FileSystem.setLength( (int)fd, (int)len );
  }

  /**
    * Tests a file descriptor for validity
    *
    * @param fd The native file descriptor
    *
    * @return <code>true</code> if the fd is valid, <code>false</code> 
    * otherwise
    *
    * @exception IOException If an error occurs
    */
  static boolean nativeValid(long fd) throws IOException {
      return VM_FileSystem.isValidFD( (int)fd );
  }

  /**
    * Flushes any buffered contents to disk
    *
    * @param fd The native file descriptor
    *
    * @exception IOException If an error occurs
    */
  static void nativeSync(long fd) throws SyncFailedException {
      VM_FileSystem.sync( (int)fd );
  }

}


