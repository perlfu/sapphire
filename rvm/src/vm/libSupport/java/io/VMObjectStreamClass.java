/*
 * (C) Copyright IBM Corp 2002
 */
//$Id$
package java.io;

import com.ibm.JikesRVM.classloader.VM_Class;
import com.ibm.JikesRVM.classloader.VM_Type;

/**
 * java.io.ObjectStream helper implemented for Jikes RVM.
 *
 * @author Dave Grove
 */
final class VMObjectStreamClass {

  static boolean hasClassInitializer (Class cls) {
    VM_Type t = java.lang.JikesRVMSupport.getTypeForClass(cls);
    if (t.isClassType()) {
      return t.asClass().getClassInitializerMethod() != null;
    } else {
      return false;
    }
  }
}
