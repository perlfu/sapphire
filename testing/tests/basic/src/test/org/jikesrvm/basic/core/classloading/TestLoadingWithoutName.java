/*
 * This file is part of Jikes RVM (http://jikesrvm.sourceforge.net).
 * The Jikes RVM project is distributed under the Common Public License (CPL).
 * A copy of the license is included in the distribution, and is also
 * available at http://www.opensource.org/licenses/cpl1.0.php
 *
 * (C) Copyright IBM Corp. 2001
 */
package test.org.jikesrvm.basic.core.classloading;

import java.io.IOException;
import java.io.InputStream;

public class TestLoadingWithoutName
    extends ClassLoader {

  public static void main(String[] args) throws IOException {

    final String resource = "TestClassLoading.class";
    System.out.println("Loading resource " + resource);
    final InputStream input = TestLoadingWithoutName.class.getResourceAsStream(resource);
    System.out.println("Loaded resource? = " + (null != input)); 
    final int size = input.available();
    final byte[] data = new byte[size];
    int count = 0;
    while (count < data.length) {
      count += input.read(data,count,data.length - count);
    }

    final TestLoadingWithoutName loader = new TestLoadingWithoutName();
    System.out.println("Defining class");
    final Class<?> clazz = loader.defineClass(null, data, 0, data.length);
    System.out.println("Class defined Expected: test.org.jikesrvm.basic.core.TestClassLoading Actual: " + clazz.getName() );
  }
}
