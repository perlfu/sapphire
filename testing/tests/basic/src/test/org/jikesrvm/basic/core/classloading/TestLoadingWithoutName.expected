Loading resource TestClassLoading.class
Loaded resource? = true
Defining class
Class defined Expected: test.org.jikesrvm.basic.core.TestClassLoading Actual: test.org.jikesrvm.basic.core.classloading.TestClassLoading
